### DISCUSSION TITLE
how to pipeline queries?

### DISCUSSION THREAD

**Discussion by davepacheco** [diesel-rs/diesel#3147] [Q&A] [answered]
I want my application to be able to send a batch of queries like:

```sql
BEGIN;
UPDATE TableA ... ;
INSERT INTO TableB ... ON CONFLICT (...) DO NOTHING RETURNING ... ;
INSERT INTO TableC ... ON CONFLICT (...) DO NOTHING RETURNING ... ;
SELECT ... ;
COMMIT;
```

I want to be able to get the count of rows updated by the UPDATE and the rows or errors returned by the INSERTs and SELECTs.  I believe the PostgreSQL wire protocol supports this, and tokio-postgres seems to support it via [pipelining](https://docs.rs/tokio-postgres/latest/tokio_postgres/#pipelining).  With Diesel I think I can use [`batch_execute()`](https://docs.rs/diesel/latest/diesel/connection/trait.SimpleConnection.html#tymethod.batch_execute) if I don't need the results of each query, and I can use [`transaction()`](https://docs.rs/diesel/latest/diesel/connection/trait.Connection.html#method.transaction) if I don't mind paying multiple client round-trips (potentially with database locks held).  Is there a way with Diesel to send all these statements as a batch _and_ get their results?

**Comment by weiznich** [comment, marked as answer]
Diesel does currently not support pipelining. We do not have any concrete planes to support this yet. Additionally this feature seems to fit better into `diesel_async` as in diesel itself, as it basically requires using either `tokio-diesel` (which is async) or the libpq async interface. See https://github.com/weiznich/diesel_async/discussions/4 for relevant discussion at the other repository. As mentioned there supporting pipelining would require some sort of abstraction that allows to support non-pipelining connection implementations as well.

**Reply by davepacheco** [reply]
Thanks for clarifying.

**Comment by Ten0** [comment]
Sooo following up on [this discussion](https://www.reddit.com/r/rust/comments/xpo2dd/comment/iqh7aiu) in this more appropriate place.
With libpq 14 adding some form of pipelining support, it may be relevant to provide such interfaces, and doesn't seem completely impossible.
This kind of interface seems to allow drastically reducing the number of round trips to the database, improving latency a lot.

Notable points:
- There are two types of pipelining:
  1. one where if one statement of the batch fails, all subsequent statements are not executed
  2. one where all statements are executed independently, and may error out independently
- Postgres supports both ([via the sync messages](https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-PIPELINING)), but mysql only supports the latter (it has no such thing as sync messages to specify the end of a batch that could be omitted - each query has implicitly a sync message behind it.)

Type 1 covers very well for cases as the original post suggests, and a nice interface for it that would allow to minimize unnecessary buffer flushes as we can omit the sync message might be one mentioned in the previous discussion:
```rust
impl PgConnection {
    fn pipeline_queries<T: QueryListForPipelining>(&mut self, t: T) -> QueryResult<T::Res>;
}
```

However this interface doesn't work as well for more complex query graphs where queries depend on results of other queries:
```mermaid
graph TD;
    A-->C;
    B-->D;
    A-->D;
    C-->D;
```
In this case you want to:
- Start A and B
- As soon as you got A's result, start C
- As soon as you got B's result and C's result, start D
- Wait for D's result

and I guess ideally would like to write it like this:
```rust
let connection: &mut PgConnection = ...;
let pl: Pipeline<'_> = connection.pipeline(); // Borrows connection
let a = query_a.pipeline(&pl)?;
let b = query_b.pipeline(&pl)?;
let a_res = a.get_results()?;
let c = query_c(&a_res).pipeline(&pl)?;
let b_res = b.get_results()?;
let c_res = c.get_results()?;
let d_res = query_d(&b_res, &c_res).get_results(connection)?;
```
Notice how a, b and c are kind-of similar to futures in that they represent the future completion of something, but it looks like they don't need to be actual `futures::Future`.

This is something that could be implemented with pipelining type 2 (on both PG and MySQL).
Compared to the `diesel-async` implementation, this interface would ideally not require that the futures are polled together for the queries to be processed pipelined.

An alternate interface could be something along those lines:
```rust
let connection: &mut PgConnection = ...;
let pl: Pipeline<'_, ()> = connection.pipeline(); // Borrows connection mutably
let pl: Pipeline<'_, (A_OUT,)> = query_a.pipeline(pl)?;
let pl: Pipeline<'_, (A_OUT, B_OUT)> = query_b.pipeline(pl)?;
let (a_res, pl): (_, Pipeline<'_, (B_OUT,)>) = pl.get_results()?;
let pl: Pipeline<'_, (B_OUT, C_OUT)> = query_c(&a_res).pipeline(pl)?;
let (b_res, pl): (_, Pipeline<'_, (C_OUT,)>) = pl.get_results()?;
let (c_res, pl): (_, Pipeline<'_, ()>) = pl.get_results()?;
let d_res = query_d(&b_res, &c_res).get_results(connection)?;
```

Versus between these two:
A:
 1. Can be dynamically managed
 2. Less risk to mess-up matching the queries and their results if several queries have the same output type. (In complex call graphs that quickly becomes very important)

B:
 1. Ensures that the queries are pulled out of the pipeline in the same order as they get in. The other interface requires:
   a. Buffers to manage storing query results when e.g. calling `get_results`  on B before A,
   b. Great attention of the user if they wish to avoid pulling results in a different order than they put queries in, which would be impactful for performance (e.g. row-by-row would be nullified if A was a query that you wanted to process in row-by-row mode but you pull B before).

I couldn't think of an interface that has B.1 (prevents accidentally pulling in incorrect order, no requirement to manage that potential incorrect-order-pulling), while conserving A.2 (the ergonomics and safety of actually-working of A).
Overall over these two I feel A would make for a better interface.

NB: I feel like for 90% of the use-cases, erroring that previous query in the pipeline has failed is correct, because we would normally not attempt them if processing them sequentially (`.load()?`), so we may still want to use mode 1 with this interface by default...

Side note: There's [this Postgres patch](https://commitfest.postgresql.org/40/3938/) that fixes a bug in this area, we might want to wait for it to be released.

**Reply by weiznich** [reply]
Thanks for the great writeup. That's a really good overview of possible options. It seems feasible to add this to diesel. As for the possible options: The first variant looks much nicer to be used, as it does not involve complex types. As you noticed that comes at the cost that user could "poll" the results in the wrong order, which might be a serve issue. On the other hand I really want to avoid adding another case of overcomplex types to diesels public API. In the end we need to experiment and see if it's possible to somehow workaround one of the both issues.
